src/main/java/org/inferred/freebuilder/processor/excerpt/CheckedBiMap.java

```diff
@@ -0,0 +1,265 @@
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public V forcePut(K key, V value) {")
+        .addLine("    V oldValue = biMap.get(key);")
+        .addLine("    put.accept(key, value);")
+        .addLine("    return oldValue;")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public V remove(Object key) {")
+        .addLine("    return biMap.remove(key);")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public void clear() {")
+        .addLine("    biMap.clear();")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public %s<%s<K, V>> entrySet() {",
+            Set.class, Map.Entry.class
+        )
+        .addLine("    return new %s<>(biMap.entrySet(), put);", CheckedBiMap.CheckedEntrySet.TYPE)
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public %s<V,K> inverse() {", BiMap.class)
+        .addLine("    return new %s<V, K>(biMap.inverse(), "
+            + "(value, key) -> put.accept(key, value));",
+            TYPE
+        )
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public %s<V> values() {", Set.class)
+        .addLine("    return biMap.values();")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public void putAll(%s<? extends K, ? extends V> map) {", Map.class)
+        .addLine("    for (%s<? extends K, ? extends V> entry : map.entrySet()) {", Map.Entry.class)
+        .addLine("      put(entry.getKey(), entry.getValue());")
+        .addLine("    }")
+        .addLine("  }")
+        .addLine("}");
+  }
+
+  @Override
+  protected void addFields(FieldReceiver fields) {
+  }
+
+  private static class CheckedEntry extends ValueType implements Excerpt {
+
+    static final LazyName TYPE = LazyName.of("CheckedEntry", new CheckedBiMap.CheckedEntry());
+
+    private CheckedEntry() {
+    }
+
+    @Override
+    public void addTo(SourceBuilder code) {
+      code.addLine("")
+          .addLine("private static class %s<K, V> implements %s<K, V> {", TYPE, Map.Entry.class)
+          .addLine("")
+          .addLine("  private final K key;")
+          .addLine("  private V value;")
+          .addLine("  private final %s<K, V> put;", BiConsumer.class)
+          .addLine("")
+          .addLine("  %s(%s<K, V> entry, %s<K, V> put) {", TYPE, Map.Entry.class, BiConsumer.class)
+          .addLine("    this.key = entry.getKey();")
+          .addLine("    this.value = entry.getValue();")
+          .addLine("    this.put = put;")
+          .addLine("  }")
+          .addLine("")
+          .addLine("  @Override public K getKey() {")
+          .addLine("    return key;")
+          .addLine("  }")
+          .addLine("")
+          .addLine("  @Override public V getValue() {")
+          .addLine("    return value;")
+          .addLine("  }")
+          .addLine("")
+          .addLine("  @Override public V setValue(V value) {")
```

Comment:
Doesn't this method need to check for collisions?

Reply:
I've added a `hasValue` predicate to pass around to detect collisions. I'm open to suggestions, that was just the simplest one that came to my mind first.

Reply:
I'd pass in the BiMap and do `biMap.containsValue(value)`, same as the put method above

Reply:
Sounds good, I've updated. That was my other thought as well. I wasn't sure which direction you wanted. Thanks.